pkg/tasks/worker/worker.go

```diff
@@ -0,0 +1,134 @@
+package worker
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	m "github.com/content-services/content-sources-backend/pkg/instrumentation"
+	"github.com/content-services/content-sources-backend/pkg/tasks/queue"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog"
+)
+
+type worker struct {
+	queue     queue.Queue
+	logger    *zerolog.Logger
+	workerWg  *sync.WaitGroup // wait for worker loop to exit
+	handlers  map[string]TaskHandler
+	taskTypes []string
+	metrics   *m.Metrics
+
+	stopFunc          context.CancelFunc // called when stop() is called to gracefully exit worker
+	readyChan         chan struct{}      // receives value when worker is ready for new task
+	heartbeatInterval time.Duration
+}
+
+type workerConfig struct {
+	queue     queue.Queue
+	logger    *zerolog.Logger
+	workerWg  *sync.WaitGroup
+	handlers  map[string]TaskHandler
+	taskTypes []string
+}
+
+func newWorker(config workerConfig, metrics *m.Metrics) worker {
+	return worker{
+		queue:     config.queue,
+		logger:    config.logger,
+		workerWg:  config.workerWg,
+		handlers:  config.handlers,
+		taskTypes: config.taskTypes,
+		readyChan: make(chan struct{}, 1),
+		metrics:   metrics,
+	}
+}
+
+func (w *worker) start(ctx context.Context, cancel context.CancelFunc) {
+	w.logger.Info().Msg("Starting worker")
+	defer w.workerWg.Done()
+	defer recoverOnPanic(w.logger)
+
+	var taskId uuid.UUID
+	var taskToken uuid.UUID
+	w.stopFunc = cancel
+	w.readyChan <- struct{}{}
+
+	beat := time.NewTimer(w.heartbeatInterval / 3)
+	defer beat.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			if taskId != uuid.Nil {
+				err := w.requeue(taskId)
+				if err != nil {
+					w.logger.Warn().Msg(fmt.Sprintf("error requeueing task: %v", err))
+				}
+			}
+			return
+		case <-w.readyChan:
+			taskInfo, err := w.fetch(ctx)
+			if err != nil {
+				if err == queue.ErrContextCanceled {
+					continue
+				}
+				w.logger.Warn().Msg(fmt.Sprintf("error dequeuing task: %v", err))
+				continue
+			}
+			if taskInfo != nil {
+				w.metrics.RecordMessageLatency(taskInfo.Queued)
+				taskId = taskInfo.Id
+				taskToken = taskInfo.Token
+				go w.process(ctx, taskInfo)
+			}
+		case <-beat.C:
+			w.queue.RefreshHeartbeat(taskToken)
```

Comment:
You'll see here the worker is periodically refreshing the task's heartbeat. A drawback is this assumes the task handler never hangs. Because the task handler is run in a goroutine, if the task handler hangs, the worker will keep refreshing the heartbeat, but the task will never finish. My thought is, since we're writing the task handlers, we might be able to make this assumption.

If we don't want to assume that the task handler never hangs,  I think the alternative to this is having the task handler refresh the task's heartbeat. In this case, we might want two heartbeats, one for the task and one for the worker (so we also know the worker isn't hanging).

Not sure.

Reply:
maybe we should chat about this in person at standup